### PR TITLE
fix(DB/SAI):Twilight Discipline and Twilight Thug scripts(talks an…

### PR DESCRIPTION
**Pull request by Ragelezz**
…d spells)

## Changes Proposed:
-  
-  

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/7963

## Tests Performed:

- .go c id 2339, no spells, no texts when engaged or died
- 

## How to Test the Changes:

1. .go c id 2339 he has talks (entering combat, dies) and spells (heal, renew on discipline and disarm on thug)
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1632248054324754400.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1632248054324754400');
+
+UPDATE creature_template SET AIName='' WHERE entry IN (2338,2339);
+UPDATE creature_template SET ScriptName='npc_twilight_discipline_thug' WHERE entry IN (2338,2339);
```
Comment (Helias) [inline data/sql/updates/pending_db_world/rev_1632248054324754400.sql:4]:
```suggestion
UPDATE `creature_template` SET `AIName`='' WHERE `entry` IN (2338,2339);
UPDATE `creature_template` SET `ScriptName`='npc_twilight_discipline_thug' WHERE `entry` IN (2338,2339);
```

File: data/sql/updates/pending_db_world/rev_1632338572160003700.sql [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES (2339, 1, 1, 'My lords! I come to thee!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Defeat');
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES (2339, 1, 2, 'You will never defeat us!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug- Defeat');
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES (2339, 1, 3, 'My life for my masters. Rejoice!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug- Defeat');
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES (2339, 1, 4, 'Your victory here only speeds your doom!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug- Defeat');
```
Comment (Helias) [inline data/sql/updates/pending_db_world/rev_1632338572160003700.sql:28]:
```suggestion
INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
(2339, 0, 0, 'Embrace the end!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Engage'),
(2339, 0, 1, 'For our Masters!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Engage'),
(2339, 0, 2, 'Glory and blood for the Old Gods!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Engage'),
(2339, 0, 3, 'Intruders! Good, I was getting bored...', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Engage'),
(2339, 0, 4, 'Intruders! Slay them!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Engage'),
(2338, 0, 0, 'Embrace the end!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Engage'),
(2338, 0, 1, 'For our Masters!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Engage'),
(2338, 0, 2, 'Glory and blood for the Old Gods!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Engage'),
(2338, 0, 3, 'Intruders! Good, I was getting bored...', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Engage'),
(2338, 0, 4, 'Intruders! Slay them!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Engage'),
(2338, 1, 0, 'My death matters little... the Hammer will still fall !', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Defeat'),
(2338, 1, 1, 'My lords! I come to thee!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Defeat'),
(2338, 1, 2, 'You will never defeat us!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Defeat'),
(2338, 1, 3, 'My life for my masters. Rejoice!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Defeat'),
(2338, 1, 4, 'Your victory here only speeds your doom!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Discipline - Defeat'),
(2339, 1, 0, 'My death matters little... the Hammer will still fall !', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Defeat'),
(2339, 1, 1, 'My lords! I come to thee!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug - Defeat'),
(2339, 1, 2, 'You will never defeat us!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug- Defeat'),
(2339, 1, 3, 'My life for my masters. Rejoice!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug- Defeat'),
(2339, 1, 4, 'Your victory here only speeds your doom!', 12, 0, 100, 0, 0, 0, 0, 0, 'Twilight Thug- Defeat');
```
https://www.azerothcore.org/wiki/sql-standards

**Review by Helias — APPROVED** [review]

File: data/sql/updates/pending_db_world/rev_1632248054324754400.sql [hunk 3]
```diff
@@ -0,0 +1,5 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1632248054324754400');
+
+UPDATE `creature_template` SET `AIName`='' WHERE `entry` IN (2338,2339);
+UPDATE `creature_template` SET `ScriptName`='npc_twilight_discipline_thug' WHERE `entry` IN (2338,2339);
```
Comment (Helias) [inline data/sql/updates/pending_db_world/rev_1632248054324754400.sql:4]:
```suggestion
UPDATE `creature_template` SET `AIName`='', `ScriptName`='npc_twilight_discipline_thug'WHERE `entry` IN (2338,2339);
```

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 4]
```diff
@@ -413,10 +413,117 @@ class npc_prospector_remtravel : public CreatureScript
...
+        }
+
+        void JustDied(Unit*) override
+        {
+            Talk(TALK_DEFEAT, NULL);
+        }
+
+        void EnterCombat(Unit*) override
+        {
+            Talk(TALK_ENGAGE, NULL);
```
Comment (Kitzunu) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:474]: nullptr.

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 5]
```diff
@@ -413,10 +413,117 @@ class npc_prospector_remtravel : public CreatureScript
...
+        {
+            events.Reset();
+
+            heal = false;
+            renew = false;
+        }
+
+        void JustDied(Unit*) override
+        {
+            Talk(TALK_DEFEAT, NULL);
```
Comment (Kitzunu) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:469]: same here

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 6]
```diff
@@ -413,10 +413,117 @@ class npc_prospector_remtravel : public CreatureScript
...
+
+            if (me->GetEntry() == NPC_TWILIGHT_THUG)
+                events.ScheduleEvent(EVENT_DISARM, 30 * IN_MILLISECONDS);
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+
+            if (!UpdateVictim())
+                return;
```
Comment (Kitzunu) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:484]: brackets

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 7]
```diff
@@ -413,10 +413,117 @@ class npc_prospector_remtravel : public CreatureScript
...
+        {
+            Talk(TALK_DEFEAT, NULL);
+        }
+
+        void EnterCombat(Unit*) override
+        {
+            Talk(TALK_ENGAGE, NULL);
+
+            if (me->GetEntry() == NPC_TWILIGHT_THUG)
+                events.ScheduleEvent(EVENT_DISARM, 30 * IN_MILLISECONDS);
```
Comment (Kitzunu) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:477]: brackets

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 8]
```diff
@@ -413,10 +413,117 @@ class npc_prospector_remtravel : public CreatureScript
...
+class npc_twilight_discipline_thug : public CreatureScript
+{
+public:
+    npc_twilight_discipline_thug() : CreatureScript("npc_twilight_discipline_thug") {}
+
+    struct npc_twilight_discipline_thug_AI : public ScriptedAI
+    {
+        npc_twilight_discipline_thug_AI(Creature* creature) : ScriptedAI(creature)
+        {
+        }
```
Comment (Kitzunu) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:453]: keep these on the same line
```suggestion
        npc_twilight_discipline_thug_AI(Creature* creature) : ScriptedAI(creature) { }
```

**Review by Kitzunu — CHANGES_REQUESTED** [review]

**Review by IntelligentQuantum — APPROVED** [review]

**Comment by Ragelezz** [comment]
ignore that many red lines, just the local core..

**Comment by temperrr** [comment]
- [x] Heal gets casted when nearly dead
- [ ] Renew gets casted just before they die (like at 5% hp), this should be earlier.
- [x] They usually say something when they engage or die.
- [x] Thugs use Disarm.

**Comment by Nyeriah** [comment]
Hello,

I've identified a few issues in this PR that need to be addressed:

1) All of those `creature_text` lines are already in the database. Could you provide some informaton on why reinserting them was required? 
Skärmbild 2021-10-30 115334

Also, you seem to have forgotten to assign their respective  `BroadcastTextId`, so translations for them will stop working after reinserted.

2) If I understood this correctly, we are reusing the same script-body to script two distinct creatures that do distinct things, only running specific code through IF checks, like here: https://github.com/azerothcore/azerothcore-wotlk/pull/8023/files#diff-e6b2ea879c413a6d5115c9b701b49aff29a19f15e058965b2b5d09faa6d29aedR504

Clustering scripts like this is not optimal, unless they are very generic creatures that perform the very same actions we shouldn't opt to stack all of them in the same script like this, it seems rather arbitrary to me and may lead to issues maintaining the code longterm (like scripts that not exactly address what they're expected to).

3) The changes proprosed here could be addressed through just a few SAI lines. While looking into this, I also found that their abilities were already scripted, so we just need to add their text on aggro. I've been using Keira:

Skärmbild 2021-10-30 120842

Once we select a creature, it offers us a very simple form that we can edit to add new SAI rows, from there we can script the line to make them talk by edding the values accordingly:

Skärmbild 2021-10-30 123833

So the following SQL is all that we really need in order to achieve our goal here:
```sql
DELETE FROM `smart_scripts` WHERE (`entryorguid` = 2338) AND (`source_type` = 0) AND (`id` IN (2, 3));
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(2338, 0, 2, 0, 4, 0, 100, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 'Twilight Disciple - On Aggro - Talk 0'),
(2338, 0, 3, 0, 6, 0, 100, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Twilight Disciple - On Death - Talk 1');

DELETE FROM `smart_scripts` WHERE (`entryorguid` = 2339) AND (`source_type` = 0) AND (`id` IN (3, 4));
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(2339, 0, 3, 0, 4, 0, 100, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 'Twilight Thug - On Aggro - Talk 0'),
(2339, 0, 4, 0, 6, 0, 100, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'Twilight Thug - On Death - Talk 1');
```

https://user-images.githubusercontent.com/47818697/139540073-3a6a1f5b-085c-4287-8747-7d61d5d89a17.mp4

**Comment by Ragelezz** [comment]
So just remove C++ and add the SQL lines provided?

**Comment by Nyeriah** [comment]
> So just remove C++ and add the SQL lines provided?

Yes, that could be done

**Comment by Helias** [comment]
There are some merge conflicts @Ragelezz

**Comment by Ragelezz** [comment]
What merge conflicts? I removed my c++ script and left only what Nyeriah posted. :-?

**Comment by Helias** [comment]
as git says, there is something wrong with `src/server/scripts/Kalimdor/zone_darkshore.cpp`

in your local machine I recommend to:
- update your fork master branch
- go to the branch of this PR
- merge master, so use `git merge origin/master`
- this will merge the master branch with the PR branch and it will show you the conflict file locally

**Comment by Ragelezz** [comment]
Hmm, strange; will have a look

**Comment by Helias** [comment]
What is the status of this?

**Comment by Ragelezz** [comment]
Guess its needing tests since review its not the case anymore,

**Comment by Nyeriah** [comment]
Testing is not required as it's using the same queries I used to make my video, so that part should be covered. But this PR got polluted with some C++ changes in the darkshore file, I suppose it could be merged as it is if they are nothing but style changes, but it's still pretty strange for them to be there

**Comment by FrancescoBorzi** [comment]
@Ragelezz as a side note, this is not a "Core" change but "Scripts" instead, as everything under `src/scripts` :)

**Comment by Nyeriah** [comment]
> @Ragelezz as a side note, this is not a "Core" change but "Scripts" instead, as everything under `src/server/scripts` :)

Let's not get confused here: the SQL in this PR is the only part that matters, the C++ is just indentation changes not really related to this PR

**Pull request merged**